akka-http/src/main/scala/akka/http/javadsl/server/Unmarshaller.scala

```diff
@@ -4,11 +4,95 @@
-  def classTag: ClassTag[T]
+object Unmarshaller {
+  private implicit val _ = javaToScalaHttpEntity
+  
+  def wrap[A,B](scalaUnmarshaller: unmarshalling.Unmarshaller[A,B]) = Unmarshaller()(scalaUnmarshaller)
+
+  /**
+   * Creates an unmarshaller from an asynchronous Java function. 
+   */
+  def async[A,B](f: java.util.function.Function[A,CompletionStage[B]]) = wrap(unmarshalling.Unmarshaller[A,B] {
+    ctx => a => f.apply(a).toScala
+  })
+  
+  /**
+   * Creates an unmarshaller from a Java function.
+   */
+  def sync[A,B](f: java.util.function.Function[A,B]) = wrap(unmarshalling.Unmarshaller[A,B] {
+    ctx => a => scala.concurrent.Future.successful(f.apply(a))
+  })
+  
+  def entityToByteString = wrapFromHttpEntity(unmarshalling.Unmarshaller.byteStringUnmarshaller)
+  def entityToByteArray = wrapFromHttpEntity(unmarshalling.Unmarshaller.byteArrayUnmarshaller)
+  def entityToCharArray = wrapFromHttpEntity(unmarshalling.Unmarshaller.charArrayUnmarshaller)
+  def entityToString = wrapFromHttpEntity(unmarshalling.Unmarshaller.stringUnmarshaller)
+  def entityToUrlEncodedFormData = wrapFromHttpEntity(unmarshalling.Unmarshaller.defaultUrlEncodedFormDataUnmarshaller)
+  
+  def requestToEntity = wrap(unmarshalling.Unmarshaller[HttpRequest,RequestEntity] {
+    ctx => request => scala.concurrent.Future.successful(request.entity())
+  })
+  
+  def forMediaType[B](t: MediaType, um:Unmarshaller[_ <: HttpEntity,B]): Unmarshaller[HttpEntity,B] = {
+    wrap(assumeScala(um.unwrap).forContentTypes(t: scaladsl.model.MediaType))
+  }
+  
+  def forMediaTypes[B](types: java.lang.Iterable[MediaType], um:Unmarshaller[_ <: HttpEntity,B]): Unmarshaller[HttpEntity,B] = {
+    wrap(assumeScala(um.unwrap).forContentTypes(types.asScala.toSeq.map(ContentTypeRange(_)): _*))
+  }
+  
+  def firstOf[A, B] (u1: Unmarshaller[A, B], u2: Unmarshaller[A, B]): Unmarshaller[A, B] = {
+    wrap(unmarshalling.Unmarshaller.firstOf(u1.unwrap, u2.unwrap))
+  }
+  
+  def firstOf[A, B] (u1: Unmarshaller[A, B], u2: Unmarshaller[A, B], u3: Unmarshaller[A, B]): Unmarshaller[A, B] = {
+    wrap(unmarshalling.Unmarshaller.firstOf(u1.unwrap, u2.unwrap, u3.unwrap))
+  }
+  
+  def firstOf[A, B] (u1: Unmarshaller[A, B], u2: Unmarshaller[A, B], u3: Unmarshaller[A, B], u4: Unmarshaller[A, B]): Unmarshaller[A, B] = {
+    wrap(unmarshalling.Unmarshaller.firstOf(u1.unwrap, u2.unwrap, u3.unwrap, u4.unwrap))
+  }
+  
+  def firstOf[A, B] (u1: Unmarshaller[A, B], u2: Unmarshaller[A, B], u3: Unmarshaller[A, B], u4: Unmarshaller[A, B], u5: Unmarshaller[A, B]): Unmarshaller[A, B] = {
+    wrap(unmarshalling.Unmarshaller.firstOf(u1.unwrap, u2.unwrap, u3.unwrap, u4.unwrap, u5.unwrap))
+  }
+  
+  private def wrapFromHttpEntity[B](scalaUnmarshaller: unmarshalling.Unmarshaller[scaladsl.model.HttpEntity,B]) = {
+    wrap(scalaUnmarshaller: unmarshalling.Unmarshaller[HttpEntity,B])
+  }
+    
+}
+
+case class Unmarshaller[A,B] private (implicit val unwrap: unmarshalling.Unmarshaller[A,B]) {
```

Comment:
added comment on why it's implicit. Also turned into `final class` instead of `case class`, and renamed `unwrap` to `asScala`.